Add Rescore functionality

peermuellerxw opened the pull request:
Closes #1686

<!--

Thank you for proposing a pull request. This template will guide you through the essential steps necessary for a pull request.
Make sure that:

-->

- [ x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [ x] There is a ticket in the bug tracker for the project in our [issue tracker](https://github.com/spring-projects/spring-data-elasticsearch/issues).
- [ x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [ x] You submit test cases (unit or integration tests) that back your changes.
- [ x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

pivotal-issuemaster commented:
@peermuellerxw Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-elasticsearch&pullRequestId=1688)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-elasticsearch&pullRequestId=1688) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@peermuellerxw Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-elasticsearch&pullRequestId=1688)!

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/query/RescorerQuery.java:
```diff
@@ -0,0 +1,40 @@
+
+	private final QueryRescorerBuilder queryRescorerBuilder;
+
+	public RescorerQuery(QueryRescorerBuilder queryRescorerBuilder) {
```
this is leaking the Elasticsearch class into the API

sothawo reviewed (changes requested):
Thanks for this PR, great work, but I think we should take this further. It is focused on the `NativeSearchQuery`, but I think we should consider the other `Query` implementations here as well.

Something like having a 
```java
public class RescorerQuery {
  Query query; // <-  our Query interface
  ScoreMode scoreMode;
  int windowSize;
}
```
not using Elasticsearch classes here, but also having our own enum for the ScoreMode. This would allow to not only use `NativeSearchQuery` but also for example a `StringQuery`. In the `RequestMapper` this would then be mapped to the corresponding Elasticsearch classes – there are existing methods to convert a `Query` to a `QueryBuilder`.

What do you think about this change?

peermuellerxw commented:
Thanks for looking into my PR and for the feedback. Your suggestions do make sense and I am looking to implement them in the following days :)

peermuellerxw commented:
I updated the PR and changed the `RescorerQuery` to a "proper" domain object. Let me know if this is what you had in mind and also if you have any other comments/suggestions. @sothawo

sothawo commented on src/main/java/org/springframework/data/elasticsearch/core/RequestFactory.java:
```diff
@@ -1260,6 +1269,30 @@ private void prepareSort(Query query, SearchRequestBuilder searchRequestBuilder,
+
+	private QueryRescorerBuilder getQueryRescorerBuilder(RescorerQuery rescorerQuery) {
+
+		QueryRescorerBuilder builder = new QueryRescorerBuilder(Objects.requireNonNull(getQuery(rescorerQuery.getQuery())));
```
Instead of a `NullPointerException` we should throw an `IllegalArgumentException` here

sothawo reviewed (approved):
Looks fine, there are just some minor points with null checks. No need for you to fix these, I will do this in a follow-up commit after merging this PR.
